torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+            mod_reduced = base.subs(self._var_to_val) % divisor.subs(self._var_to_val)
+            congruence = (base - mod_reduced) % divisor
+            if congruence != 0:
+                self._congruences[s].add(congruence)
+            return mod_reduced
+
+        def floor_div_handler(*args):
+            # Suppose that we have an expression of the form b // d with free variable s.
+            # Using the value of s, we can evaluate b % d to a value k.
+            # Then we can rewrite b // d to (b - k) / d, while adding the guard b % d == k.
+
+            # NOTE(avik): This is exactly equivalent to rewriting b // d as (b - (b % d)) / d
+            # and eliminating b % d as above.
+            base, divisor = args
+            base, divisor = self.rewrite_with_congruences(s, base), self.rewrite_with_congruences(s, divisor)
+            mod_reduced = base.subs(self._var_to_val) % divisor.subs(self._var_to_val)
+            congruence = (base - mod_reduced) % divisor
+            if congruence != 0:
+                self._congruences[s].add(congruence)
+            return (base - mod_reduced) / divisor
+
+        if expr.has(sympy.Mod):
+            expr = expr.replace(sympy.Mod, mod_handler)
+        if expr.has(FloorDiv):
+            expr = expr.replace(FloorDiv, floor_div_handler)
+        return expr
+
+    def add(self, expr):
+        if expr == True:
+            return
+        orig_expr = expr
+        orig_reduced = orig_expr.subs(self._var_to_val)
+        assert orig_reduced != False, f"{orig_expr} is inconsistent!"
+        free_symbols = expr.free_symbols
+        assert free_symbols, f"Did not expect constraint with no free variables: {expr}"
+        if len(free_symbols) > 1:
+            # multivariate: record and move on
+            self._multivariate_inequalities.add(expr)
+        else:
+            # univariate: can solve these immediately
+            s = next(iter(free_symbols))
+            # eliminate // and % (see documentation of `rewrite_with_congruences` above)
+            expr = self.rewrite_with_congruences(s, expr)
+            if expr != True:
+                reduced = expr.subs(self._var_to_val)
+                assert reduced != False, f"{expr}, obtained by rewriting {orig_expr} with congruences, is inconsistent!"
+                if isinstance(expr, sympy.Eq):
+                    # special status for symbols that have equalities (see `solve` below)
+                    self._symbols_with_equalities.add(s)
+                self._univariate_inequalities[s].add(expr)
+
+    def add_equality(self, source, expr):
+        if expr.free_symbols:
+            # these will resolve to either specializations or dynamic equality constraints
+            self._symbolic_equivalences.append((source, expr))
+        else:
+            # specialization, right here
+            self._static_results.add(f"{source.name()} == {expr}")
```

Comment:
Wait, why only here? It seems like every expr should be checked for this, right?

Reply:
Could you elaborate? We have an assertion in `add()`, we don't expect non-true non-false exprs with no free symbols.